### DISCUSSION TITLE
Best practice way to add hparams AFTER __init__ is called

### DISCUSSION THREAD

**Discussion by ssharpe42** [Lightning-AI/pytorch-lightning#11107] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
I would like to add hparams which are not args in the init method. 

These values come after running `.setup()` function to process data (essentially to get the number of levels of each categorical )
When I load the model later on I would also like these values to populate. 

Should I have these as optional parameters and update hparams after setup? Or should I be creating a DataModule and pass them in as hparams to the LightningModule?

**Comment by rohitgr7** [comment]
I think you can just manually attach it.
can you try?
```py
def setup(...):
    self.hparams['something'] = ...
```

**Reply by ssharpe42** [reply]
Ah I should have added this caveat. I need this parameter before I create the model in __init__ so I think I have to do this in the DataModule and pass it in as an argument
